Owner can delete organizations

juanignaciosl opened the pull request:
This closes #12049.

Acceptance:

- [x] In staging, with default configuration, this shouldn't appear (not even with organization-admins flag).
- [x] Delete an account works as usual: it doesn't ask for the password for Google signups, it does for normal users.

In order to test it, you must set `cartodb_com_hosted` to `true` (either in staging or in local). This is only meant for onpremise.

- [x] Admins don't see that option.
- [x] Create an organization with two users (one should be a viewer, both should have datasets imported) and share one dataset with another. Delete the org with the new option. Check that after deletion, user database and all metadata (visualizations and so on) are deleted.

javitonino commented on app/controllers/admin/organizations_controller.rb:
```diff
@@ -24,6 +26,21 @@ def show
     end
   end
 
+  def delete
```
Shouldn't this be `destroy` to match Rails conventions?

javitonino commented on lib/carto/configuration.rb:
```diff
@@ -53,6 +53,10 @@ def custom_app_views_paths
       (config && config['custom_paths'] && config['custom_paths']['views']) || Array.new
     end
 
+    def saas?
```
This was long overdue 🤗

javitonino commented on spec/models/user_spec.rb:
```diff
@@ -1455,6 +1455,34 @@ def create_role(user)
+      describe 'User#destroy' do
+        include TableSharing
+
+        it 'blocks deletion even with shared entities' do
```
I think this read better if you remove `even` from the test name

juanignaciosl replied:
_i think this riid bittir_... :D 

👍  :-)

javitonino replied:
Damn that ypirqui bot :)

javitonino commented on app/models/user.rb:
```diff
@@ -397,7 +397,7 @@ def before_destroy
       end
 
-      unless can_delete
+      unless can_delete || @force_destroy
```
Maybe `@force_destroy` should also destroy unregistered tables. But let's not do that in this PR (or in fact, until the need arises).

juanignaciosl replied:
I had the same thought. Nevertheless, as this is not meant for SaaS, I still have doubts.

javitonino replied:
Uhm, true. Well, this could be used in the SaaS, but from superadmin only. Anyway, this is not needed right now, and I prefer minimizing onpremise-only features, as they are usually harder to debug, and have a slower feedback cycle.

javitonino commented:
Left some minor comments, the only significant one is the `delete`/`destroy` naming. The PR looks nice and well tested.

I took a look at the frontend part and I didn't see anything, but it would be great if someone from @CartoDB/frontend could also take a look.

juanignaciosl commented:
@javitonino I'll review your comments when frontend people write down theirs.

xavijam commented on lib/assets/javascripts/cartodb/common/delete_organization_view.js:
```diff
@@ -0,0 +1,32 @@
+    return this.template({
+      formAction: cdb.config.prefixUrl() + '/organization',
+      authenticityToken: this.options.authenticityToken,
+      passwordNeeded: this.options.user.attributes.needs_password_confirmation === true
```
I'd try to create a local var in the initialize and then use it wherever you need.
```js
this._userModel = this.options.user;
…
passwordNeeded: !!this._userModel.get('needs_password_confirmation'),
…
```

juanignaciosl commented:
Screenshots:

![captura de pantalla 2017-05-23 a las 14 31 39](https://cloud.githubusercontent.com/assets/932968/26354377/a04158f6-3fc4-11e7-9c33-c8c5a9bd6272.png)

![captura de pantalla 2017-05-23 a las 14 31 52](https://cloud.githubusercontent.com/assets/932968/26354383/a43a5390-3fc4-11e7-84b3-3aa6d6b2a5fe.png)

xavijam commented on lib/assets/javascripts/cartodb/common/delete_organization_view.js:
```diff
@@ -20,13 +20,14 @@ module.exports = BaseDialog.extend({
   initialize: function () {
     this.elder('initialize');
     this.template = cdb.templates.getTemplate('common/views/delete_organization');
+    this.passwordNeeded = this.options.user.attributes.needs_password_confirmation === true
```
this._userModel = this.options.user;

juanignaciosl replied:
Yup, sorry.

xavijam replied:
:cabrito:

xavijam commented on lib/assets/javascripts/cartodb/common/delete_organization_view.js:
```diff
@@ -20,13 +20,14 @@ module.exports = BaseDialog.extend({
       formAction: cdb.config.prefixUrl() + '/organization',
       authenticityToken: this.options.authenticityToken,
-      passwordNeeded: this.options.user.attributes.needs_password_confirmation === true
+      passwordNeeded: this.passwordNeeded
```
passwordNeeded: !!this._userModel.get('needs_password_confirmation'),

javitonino commented on app/controllers/admin/organizations_controller.rb:
```diff
@@ -7,13 +7,15 @@ class Admin::OrganizationsController < Admin::AdminController
 
   ssl_required :show, :settings, :settings_update, :regenerate_all_api_keys, :groups, :auth, :auth_update,
-               :notifications, :new_notification, :destroy_notification
+               :notifications, :new_notification, :destroy_notification, :delete
```
You forgot to rename it in the filters in the `delete` -> `destroy` change. (more below). You would see this in tests, but I'm faster!

juanignaciosl replied:
Dude, wait! :D

javitonino replied:
Yeah, I got confused with the user refactor one, and started doing the CR, noticed this and realized that test hadn't run and I was on the wrong PR. I'll wait for you to fix this + tests before taking a second look.

Definitely a case of being assigned to too much stuff., I'll deassign myself from a bunch of stuff to avoid confusion.

javitonino commented on app/views/admin/organizations/settings.html.erb:
```diff
@@ -201,6 +201,23 @@
       </div>
     <% end %>
+
+    <% if !saas? && current_user.organization_owner? && current_user.has_feature_flag?('organization-admins')  %>
```
Ha! Durign acceptance I realized that non-owners don't even see this page, so the check for owner is not actually needed (but better to keep it, double safety).

javitonino commented:
Found a good one:
`Error deleting organization: {:user=>"Viewer users can't delete visualizations"}`

Moving back to development.

juanignaciosl commented:
@javitonino I can't reproduce, could you please provide STR? That error should be covered by [this](https://github.com/CartoDB/cartodb/blob/8c3e64e6bcea9a64053796f370cc4ed90c141a06/app/models/user.rb#L381), added time ago. I've just added a test for that, just in case.

javitonino commented:
Maybe this helps: https://rollbar.com/vizzuality/CartoDB/items/23094/
It seems it is related with that user having shared entities and that line of code (which I added with a very helpful commit message U_U). It's trying to delete shared visualizations before deleting the user. Maybe this can be removed with the new `force_destroy`?

javitonino commented on spec/models/organization_spec.rb:
```diff
@@ -78,33 +80,33 @@ def get_organization
+      ::User.where(id: user2.id).first.should be nil
       ::User.where(id: owner.id).first.should be nil
-      Carto::Visualization.exists?(vis.id).should be_false
+      Carto::UserTable.exists?(table1.id)
```
This does not test anything 😅

juanignaciosl replied:
F*cking master, I checked that debugging and pasted without adding the condition. Fixing...

Good catch, indeed.

javitonino commented:
Acceptance comments:
- We have the typical issues with unregistered tables, but that's a known limitation. After deleting the tables it continues.
- You get a bunch of Rollbar warnings about sending visualization-unshared notifications ([1](https://rollbar.com/vizzuality/CartoDB/items/23107/), [2](https://rollbar.com/vizzuality/CartoDB/items/23109/), [3](https://rollbar.com/vizzuality/CartoDB/items/23110/)). It would be nice to reduce this noise, but it's not very important as it doesn't break the flow.

I think we can launch in the current state, but maybe you want a chance to take a look at those notification errors.

juanignaciosl commented:
@javitonino first is known (#11654), second is not too relevant. I'd deliver this now, open an issue about the second, and prioritize both.
